src/main/java/com/puppycrawl/tools/checkstyle/api/FileText.java

```diff
@@ -78,14 +78,14 @@
     private final Charset charset;
 
     /**
-     * The full text contents of the file.
+     * The lines of the file, without terminators.
      */
-    private final String fullText;
+    private final String[] lines;
```

Comment:
If whole problem is because this field is final, let's remove final from it.

It looks very weird that we are damaging our API, in favor of some testing tool, because tool have a bug.

It is better even sacrifice coverage, but do not damage API.

Please explain selected approach, as I might miss something.

Reply:
@romani That's what is happening. `final` is removed from `fullText` and the field order is changed in order to keep final and non-final fields grouped. The diff is a bit confusing.

Reply:
Probably I was not clear in my comment.

In fix for this issue, should be NO changes in API ( no new methods). If problem is in final modificator let's remove it in all required places. 
As last resort, we sacrifice coverage rate.

Removal of final might not like some other tool, we will suppress violations

Reply:
I don't know whether a new package-protected method can be called a change in the API. That's the purpose of the scope package-protected to allow adding such methods, make them available for classes in the same package, including tests. I don't understand your use of the word "damage" in your previous comment. What will be damaged by this change? I'd appreciate an example. I do not understand the sentence

> If problem is in final modificator let's remove it in all required places.

The modifier has to be removed in one place only.

I don't want to push this change. I'm convinced that it has to be done in order to be able to build on OpenJDK >= 12, though. I don't see any alternative for it so far, but I have to admit that I don't get your point. I hope you find the energy to explain more. Maybe I just don't get it.

Reply:
thanks a lot for desire to understand and help to finish this PR.

Our API is not very good, full of unnecessary details and implementations and ... very long history.
We try to fix it and eventually come with pure interfaces in API package. We try our best to not create new methods in API, the less contract/api with plugins/extensions the better for us.

For protected and package scopes .... We try to practice idea that class have no friends (term from c++) at all ! That is why such scopes appear in java, friends are considered as good idea decades ago.
So class should declare his public methods and be operated only by public methods. It is very spread problem that method become "visible" for "test" code, and misused by some other code in "main" code. As result - NO friends, no exposure of non-required methods.  

That is why you can find a lot of cases in tests that we prefer reflection vs protected/package methods, and such tests are not fragile as it is usually considered. They are same ease maintainable as others. Creation of test based on reflection is also very rare case and last resort, that is explained in comment.

> Searching for /* package * / shows that you use package protected fields in the code

Not everything is good in our code base, sometime I do not have time to think about all aspects of code and discuss in PRs till code is good.

------

Issue that we try to resolve is about tests that try to use easy reflection to make 100% coverage.
If reflection approach does not work any more in jdk12 due to "final" modifiers on fields, lets remove such modifiers. We had such modifiers only because it was next to nothing to have them and it is good practice.
Lets  comment out problematic tests and see what we will get in coverage  report. We more and more become to rely on pitest for coverage, so we might sacrifice jacoco coverage partially.

You can open few more PRs to show how different versions of fix for issue.